### PR TITLE
ROS SITL

### PR DESCRIPTION
**Pull request by thomasgubler**
Bringing up support for compiling modules for NuttX and Linux/ROS from the same code. The intent is to lower the barriers for the huge pool of ROS developers for creating px4 software. For people developing robotics software this allows them to easily test the behavior first on their PC in a SITL setup with ROS/Gazebo.

**This branch is not ready for merge yet. The goal is to gather feedback and make the branch available for first testers.**

  * Main change is wrapper around the uORB and the parameter interface. On ROS, instead of uORB the standard ROS subscriber and publishers are used. The parameters map to the ROS parameter server.
    * It should be relatively straightforward to adapt this wrapper to a future uorb2
  * Similar to the ROS nodehandle there is a nodehandle for px4 which handles all subscriptions and publications of a module.
  * Example usage in the [publisher](https://github.com/PX4/Firmware/blob/ros/src/examples/publisher/publisher_example.cpp) and [subscriber](https://github.com/PX4/Firmware/blob/ros/src/examples/subscriber/subscriber_example.cpp) example
  * Note that you *don't* have to use these wrappers as long as your module should only run on a FMU/NuttX as until now. You also *don't* need to install ROS for normal PX4 development. The only major change in this case is the uorb header generation as described below.
  * Messages are now defined in the ROS msg format and then converted to uorb header files or ROS header files on compilation
    * Ported messages are here: https://github.com/PX4/Firmware/tree/ros/msg (I only ported those which are needed by one of the ROS enabled apps)
    * Regressions:
      * no support for multi-dimensional arrays. As we only used them for rotation matrices in the uorb messages this is not a big problem. Rotation matrices are now stored as a float[9] array and there is this macro for element access: https://github.com/PX4/Firmware/blob/ros/src/platforms/px4_defines.h#L208
      *  for topic groups (multiple topics of same type that are defined in same header file) I created a new message file for each topic. To make sure that the 'base' type can be used there are some manual typedefs needed: https://github.com/PX4/Firmware/blob/ros/src/modules/uORB/uORB.h#L294
      * enums are not supported: I defined the values as constants: https://github.com/PX4/Firmware/blob/ros/msg/vehicle_status.msg#L2
    * *TODO*: The message generation scripts (genmsg and gencpp from the ROS project) are included as submodules. We have to make sure they run on Mac and Windows as well.

  * Currently only the mc_att_control app is ported. The ported version can be found in [mc_att_control_multiplatform](https://github.com/PX4/Firmware/blob/ros/src/modules/mc_att_control_multiplatform/mc_att_control.cpp) (built only with the new px4fmu-v2_multiplatform make target) for now. So this PR *should* have 0 effect on regular MC flying.
  * There are some dummy ROS nodes which translate from standard ROS topics to px4 ROS topics [here](https://github.com/PX4/Firmware/tree/ros/src/platforms/ros/nodes)
  * **Setup Instructions**: https://pixhawk.org/dev/ros/sitl

Credits also to @tumbili (all the mathlib porting, initial simulation setup, ...) and @LorenzMeier 

The PR includes #1040 by @jgoppert without the changes to mc_att_control

Links to 'hacks' which should be sorted out somehow:
  * https://github.com/PX4/Firmware/blob/ros/src/modules/uORB/uORB.h#L291
  * https://github.com/PX4/Firmware/blob/ros/src/platforms/px4_defines.h#L181

### TODO Documentation update on merge
  * Toolchain setup: install python-empy

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Awesomeness! Looking forward to play with this.

**Comment by thomasgubler** [comment]
Message generation is verified on OS X

**Comment by AndreasAntener** [comment]
FYI: for message header generation under OS X, "empy" might be missing (was the case for me)
```sudo pip install -U empy```

**Comment by LorenzMeier** [comment]
@thomasgubler Could you merge the position control work into this branch and rebase on master?

**Comment by jgoppert** [comment]
Nice!

File: Makefile [hunk 1]
```diff
@@ -104,13 +104,13 @@ DESIRED_FIRMWARES 	 = $(foreach config,$(CONFIGS),$(IMAGE_DIR)$(config).px4)
 STAGED_FIRMWARES	 = $(foreach config,$(KNOWN_CONFIGS),$(IMAGE_DIR)$(config).px4)
 FIRMWARES		 = $(foreach config,$(KNOWN_CONFIGS),$(BUILD_DIR)$(config).build/firmware.px4)
 
-all:			checksubmodules $(DESIRED_FIRMWARES)
+all:	$(DESIRED_FIRMWARES)
```
Comment (jgoppert) [inline Makefile]: Was this checksubmodules deletion intended?

**Pull request merged**